sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,218 @@
-  const az_span_seq token_seq = az_build_header_callback(&header);
-  char * str_header;
-  AZ_RETURN_IF_FAILED(az_span_seq_to_new_str(token_seq, &str_header));
+  // set URL
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(p_hrb->url.size + 1);
+  char * buffer = (char *)p_writable_buffer;
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = p_hrb->url.size + 1 };
+  az_result result = az_write_url(writable_buffer, az_span_to_const_span(p_hrb->url));
+  if (az_succeeded(result)) {
+    curl_easy_setopt(p_curl->p_curl, CURLOPT_URL, buffer);
+  }
+  memset(p_writable_buffer, 0, p_hrb->url.size + 1);
+  free(buffer);
 
-  p_state->p_list = curl_slist_append(p_state->p_list, str_header);
+  // send
+  CURLcode res = curl_easy_perform(p_curl->p_curl);
+  if (res != CURLE_OK) {
+    fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+  }
+
+  // TODO: would this be part of done function instead?
+  curl_slist_free_all(headers.p_list);
 
-  free(str_header);
   return AZ_OK;
 }
 
-az_result az_build_headers(az_http_request const * const p_request, az_headers_data * p_headers) {
-  // create callback for visitor
-  az_pair_visitor const pair_visitor = az_headers_to_curl_callback(p_headers);
+/**
+ * @brief This is the function that curl will use to write response into a user provider span
+ * Function receives the size of the response and must return this same number, otherwise it is
+ * consider that function failed
+ *
+ * @param contents
+ * @param size
+ * @param nmemb
+ * @param userp
+ * @return int
+ */
+int write_to_span(void * contents, size_t size, size_t nmemb, void * userp) {
+  size_t realsize = size * nmemb + 1;
+  az_span * const mem = (az_span * const)userp;
+
+  if (mem->size == 0) {
+    mem->begin = malloc(realsize);
+    mem->size = realsize;
+  } else if (mem->size < realsize) {
+    uint8_t * ptr = realloc(mem->begin, realsize);
+    if (ptr == NULL) {
+      /* out of memory! */
+      printf("not enough memory (realloc returned NULL)\n");
+      return AZ_ERROR_ARG;
+    }
+    mem->begin = ptr;
+  }
+
+  memcpy(&(mem->begin[0]), contents, realsize);
+  mem->size += realsize;
+  mem->begin[mem->size] = 0;
+
+  // This callback needs to return the response size or curl will consider it as it failed
+  return (int)realsize - 1;
+}
+
+/**
+ * handles POST request
+ */
+az_result az_curl_post_request(
+    az_curl * const p_curl,
+    az_http_request_builder const * const p_hrb,
+    az_span * const response) {
+  (void)p_hrb;
+  // Method
+  // TODO: curl_easy_setopt(p_curl->p_curl, CURLOPT_POSTFIELDS, p_hrb->body.begin);
+
+  // URL
+  // char * url;
+  // TODO: AZ_RETURN_IF_FAILED(az_http_url_to_new_str(p_hrb, &url));
+  // curl_easy_setopt(p_curl->p_curl, CURLOPT_URL, url);
+  // free(url);
+
+  curl_easy_setopt(p_curl->p_curl, CURLOPT_WRITEFUNCTION, write_to_span);
```

Comment:
CURL error handling

Reply:
please ignore POST for now. I need request_builder to support body to enable POST method